qiskit/pulse/transforms.py

```diff
@@ -375,6 +379,85 @@ def align_sequential(schedule: Schedule) -> Schedule:
     return aligned
 
 
+def align_equispaced(schedule: Schedule,
+                     duration: int) -> Schedule:
+    """Schedule a list of pulse instructions with equivalent interval.
+
+    Args:
+        schedule: Input schedule of which top-level ``child`` nodes will be
+            reschedulued.
+        duration: Duration of context. This should be larger than the schedule duration.
+
+    Returns:
+        New schedule with input `schedule`` child schedules and instructions
+        aligned with equivalent interval.
+
+    Notes:
+        This context is convenient for writing PDD or Hahn echo sequence for example.
+    """
+    if duration and duration < schedule.duration:
+        return schedule
+    else:
+        total_delay = duration - schedule.duration
+
+    if len(schedule._children) > 1:
+        # Calculate the interval in between sub-schedules.
+        # If the duration cannot be divided by the number of sub-schedules,
+        # the modulo is appended and prepended to the input schedule.
+        interval, mod = np.divmod(total_delay, len(schedule._children) - 1)
+    else:
+        interval = 0
+        mod = total_delay
+
+    # Calculate pre schedule delay
+    delay, mod = np.divmod(mod, 2)
+
+    aligned = Schedule()
+    # Insert sub-schedules with interval
+    _t0 = int(aligned.stop_time + delay + mod)
+    for _, child in schedule._children:
+        aligned.insert(_t0, child, inplace=True)
+        _t0 = int(aligned.stop_time + interval)
+
+    return pad(aligned, aligned.channels, until=duration, inplace=True)
```

Comment:
Why do we need to pad this? Same below? The user can pad if they want.

Reply:
At least we need to insert delay from t0 to the first pulse and from the last pulse to the specified duration. I wanted to avoid implementing the logic to simplify the code and I don't find any negative effect to call the pad function here.

Reply:
Why do we need to add this delay? Shouldn't the time of the instruction be shifted instead? Then the user can decide if they need it with `pulse.pad`.

Reply:
It is also possible (and usually the delay from the last pulse to the end is zero). However user may want to apply this context to a single pulse
```python
pulse.x(d0)
with pulse.align_equispaced(duration=100):
  pulse.x(d0)
pulse.x(d0)
```
This may be an edge case but we should guarantee the returned schedule have duration of 100 in every case.

Another point I want to discuss is the definition of duration. When we write Hahn echo or other DD sequences usually we define the free induction time as the horizontal axis rather than the total pulse duration. In this case it is more convenient if we can specify the length of delay instead of duration.

Reply:
The user can always wrap this in 
```python
pulse.x(d0)
with pulse.pad():
    with pulse.align_equispaced(duration=100):
        pulse.x(d0)
pulse.x(d0)
```
I prefer to have complex functionality be composed from basic functionality, rather than hardcoded.

Reply:
Does this work? I think we need to specify `until` to `pad`, but builder context doesn't take it.

Reply:
Sure, I see now. At least it makes sense to pad the ends so we might as well pad the entire body.